framework-cloud/framework-openshift/src/test/java/org/kie/cloud/openshift/log/InstancesLogCollectorRunnableTest.java

```diff
@@ -0,0 +1,187 @@
+        ExecutorService executorService = retrieveExecutorServiceField();
+
+        cut.run();
+        cut.run();
+
+        assertEquals(1, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        cut.closeAndFlushRemainingInstanceCollectors(DEFAULT_WAIT_FOR_COMPLETION_IN_MS);
+
+        assertEquals(0, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        checkLog("BONJOUR", true);
+    }
+
+    @Test
+    public void manyInstancesRunning() {
+        setInstanceMocks("BONJOUR", "HELLO", "BUON GIORNO", "HALLO", "DOBRY DEN");
+        ExecutorService executorService = retrieveExecutorServiceField();
+
+        cut.run();
+
+        assertEquals(5, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        cut.closeAndFlushRemainingInstanceCollectors(DEFAULT_WAIT_FOR_COMPLETION_IN_MS);
+
+        assertEquals(0, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        checkLog("BONJOUR", true);
+        checkLog("HELLO", true);
+        checkLog("BUON GIORNO", true);
+        checkLog("HALLO", true);
+        checkLog("DOBRY DEN", true);
+    }
+
+    @Test
+    public void killBeforeFinished() {
+        setInstanceMocks("BONJOUR");
+        ExecutorService executorService = retrieveExecutorServiceField();
+
+        cut.run();
+
+        assertEquals(1, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        // Here we wait less than the time for the message to be delivered
+        cut.closeAndFlushRemainingInstanceCollectors(1000);
+
+        assertEquals(0, ((ThreadPoolExecutor) executorService).getActiveCount());
+
+        checkLog("BONJOUR", false);
+    }
+
+    private List<OpenShiftInstance> setInstanceMocks(String... messages) {
+        List<OpenShiftInstance> instances = Arrays.asList(messages)
+                                                  .stream()
+                                                  .map(this::createInstanceMock)
+                                                  .collect(Collectors.toList());
+
+        Mockito.when(projectMock.getAllInstances())
+               .thenReturn(instances.stream()
+                                    .map(inst -> (Instance) inst)
+                                    .collect(Collectors.toList()));
+
+        return instances;
+    }
+
+    private OpenShiftInstance createInstanceMock(String message) {
+        OpenShiftInstance instanceMock = Mockito.mock(OpenShiftInstance.class);
+        Mockito.when(instanceMock.getName()).thenReturn(message);
+
+        Mockito.when(instanceMock.observeLogs()).then((invocation) -> {
+            return Observable.fromCallable(() -> {
+                Thread.sleep(WAIT_BEFORE_MSG_IN_MS);
+                return message;
+            });
+        });
+        return instanceMock;
+    }
+
+    private ExecutorService retrieveExecutorServiceField() {
+        ExecutorService executor;
+        try {
+            Field field = cut.getClass().getDeclaredField("executorService");
```

Comment:
I would prefer to access executorService directly and not by reflection.

Reply:
Updated